addons/token-exchange/green_secret_controller.go

```diff
@@ -0,0 +1,112 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type greenSecretTokenExchangeAgentController struct {
+	hubKubeClient     kubernetes.Interface
+	hubSecretLister   corev1lister.SecretLister
+	spokeKubeClient   kubernetes.Interface
+	spokeSecretLister corev1lister.SecretLister
+	clusterName       string
+	recorder          events.Recorder
+}
+
+func newgreenSecretTokenExchangeAgentController(
+	hubKubeClient kubernetes.Interface,
+	hubSecretInformers corev1informers.SecretInformer,
+	spokeKubeClient kubernetes.Interface,
+	spokeSecretInformers corev1informers.SecretInformer,
+	clusterName string,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &greenSecretTokenExchangeAgentController{
+		hubKubeClient:     hubKubeClient,
+		hubSecretLister:   hubSecretInformers.Lister(),
+		spokeKubeClient:   spokeKubeClient,
+		spokeSecretLister: spokeSecretInformers.Lister(),
+		clusterName:       clusterName,
+		recorder:          recorder,
+	}
+	queueKeyFn := func(obj runtime.Object) string {
+		key, err := cache.MetaNamespaceKeyFunc(obj)
+		if err != nil {
+			return ""
+		}
+		return key
+	}
+
+	eventFilterFn := func(obj interface{}) bool {
+		metaObj, has := obj.(metav1.Object)
+		if !has {
+			return false
+		}
+		return metaObj.GetLabels()[common.SecretLabelTypeKey] == string(common.DestinationLabel)
+	}
+
+	hubSecretInformer := hubSecretInformers.Informer()
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(queueKeyFn, eventFilterFn, hubSecretInformer).
+		WithSync(c.sync).
+		ToController(fmt.Sprintf("%s-controller", TokenExchangeName), recorder)
+}
+
+// sync secrets with label `multicluster.odf.openshift.io/secret-type: GREEN` from hub to managed cluster
+func (c *greenSecretTokenExchangeAgentController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addon deploy %q", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore secret whose key is not in format: namespace/name
+		return nil
+	}
+
+	secret, err := getSecret(c.hubSecretLister, name, namespace)
+	if err != nil {
+		return fmt.Errorf("failed to get the sercet %q in namespace %q in hub. Error %v", name, namespace, err)
+	}
+
+	if secret.GetLabels()[common.SecretLabelTypeKey] != string(common.DestinationLabel) {
+		klog.Infof("secret %q in namespace %q is not a green secret. Skip syncing with the spoke cluster", secret.Name, namespace)
+		return nil
+	}
+
+	toNamespace := string(secret.Data["namespace"])
+	if toNamespace == "" {
+		return fmt.Errorf("missing storageCluster namespace info in secret %q in namespace %q", secret.Name, namespace)
+	}
+
+	newSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secret.Name,
+			Namespace: toNamespace,
+			Labels:    map[string]string{CreatedByLabelKey: CreatedByLabelValue},
+		},
+		Data: map[string][]byte{"secret-data": secret.Data[string(common.SecretDataKey)]},
```

Comment:
Shouldn't this be just Data: and not a key inside Data?

Reply:
We don't need to pass  the `namespace` and `storagecluster` name when coping secret from Hub to Managed cluster.

Reply:
ok. I think I understand what you mean here.

Reply:
yes i can see a problem here

![image](https://user-images.githubusercontent.com/15073590/131375403-39368def-bd94-47e3-b927-f6733c7019a1.png)

Reply:
We need a format like:
![image](https://user-images.githubusercontent.com/15073590/131375821-c136a8b0-2098-458e-ab80-90551fa9ebe8.png)

Reply:
@umangachapagain @raghavendra-talur :  I'll add a follow up PR after this one to fix this issue today.

Reply:
Ack